cmd/kubectl-gadget/utils/trace.go

```diff
@@ -290,32 +448,160 @@ RetryLoop:
-			contextLogger.Fatalf("Failed to run the gadget on all nodes: None of them succeeded")
+			return errors.New("Failed to run the gadget on all nodes: None of them succeeded")
 		}
+
 		break RetryLoop
 	}
 
+	client, err := k8sutil.NewClientsetFromConfigFlags(KubernetesConfigFlags)
+	if err != nil {
+		return fmt.Errorf("Error in creating setting up Kubernetes client: %w", err)
+	}
+
 	if customResultsDisplay == nil {
-		if outputMode != "Stream" {
-			panic(fmt.Errorf("OutputMode=%q needs a custom display function", outputMode))
+		/*
+		 * If we arrived here, it means we have at least one successful node, so it
+		 * is safe to use trace pointer.
+		 */
+		if trace.Spec.OutputMode != "Stream" {
+			return fmt.Errorf("OutputMode=%q needs a custom display function", trace.Spec.OutputMode)
 		}
-		genericStreamsDisplay(contextLogger, client, params, &results, transformLine)
+
+		params := &CommonFlags{
+			OutputMode: config.OutputMode,
+			Node:       config.NodeName,
+			Verbose:    config.Verbose,
+		}
+
+		genericStreamsDisplay(client, params, &traces, transformLine)
 	} else {
-		customResultsDisplay(contextLogger, nodes, &results)
+		nodes, err := client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("Error listing nodes: %w", err)
+		}
+
+		customResultsDisplay(nodes, &traces)
+	}
+
+	return nil
+}
+
+// DeleteTrace deletes the traces for the given trace ID using RESTClient.
+func DeleteTrace(traceID string) error {
+	traceRestClient, err := getRestClient()
+	if err != nil {
+		return err
 	}
-	deleteTraces(contextLogger, traceRestClient, traceID)
+
+	deleteTraces(traceRestClient, traceID)
+
+	return nil
+}
+
+// getTraceListFromParameters returns traces associated with the given arguments.
+func getTraceListFromParameters(config TraceConfig) ([]gadgetv1alpha1.Trace, error) {
+	namespace := ""
+	if !config.UseAllNamespaces {
+		namespace = GetNamespace()
+	}
+
+	/*
+	 * If some of the arguments are empty, they will just be ignored when
+	 * searching for corresponding trace.
+	 */
+	labels := fmt.Sprintf("GadgetName=%v,nodeName=%v,namespace=%v,podName=%v,containerName=%v,outputMode=%v",
+		config.GadgetName, config.NodeName, namespace, config.PodName, config.ContainerName, config.TraceOutputMode)
+
+	var listTracesOptions = metav1.ListOptions{
+		LabelSelector: labels,
+		FieldSelector: fields.Everything().String(),
+	}
+
+	traceRestClient, err := getRestClient()
+	if err != nil {
+		return []gadgetv1alpha1.Trace{}, err
+	}
+
+	var traces gadgetv1alpha1.TraceList
+
+	err = traceRestClient.
+		Get().
+		Namespace("gadget").
+		Resource("traces").
+		VersionedParams(&listTracesOptions, scheme.ParameterCodec).
+		Do(context.TODO()).
+		Into(&traces)
+
+	if err != nil {
+		return []gadgetv1alpha1.Trace{}, err
+	}
+	if len(traces.Items) == 0 {
+		return []gadgetv1alpha1.Trace{}, fmt.Errorf("No traces found with the given parameters %s: %w", labels, err)
+	}
```

Comment:
I don't think it's actually an error. Just return an empty list and the caller could handle it. (Otherwise it's very difficult for the caller to understand if there was an error or if there are not traces).